### DISCUSSION TITLE
VM tools are not getting installed when we try install through GO SDK

### DISCUSSION THREAD

**Discussion by venkathangella** [vmware/govmomi#3273] [Community Help]
Hi Team, 

I am using golang SDK of VMWare ("github.com/vmware/govmomi") latest version for VM tasks automation. When I tried VM tools installation using below code, it does not work as expected. Is this correct? am I missing something? When it is beyond mounting why the function name says mount installer? do we need user intervention after this?

running, err := vm.IsToolsRunning(ctx)
	if err != nil {
		return err
	}
	if running {
		return nil
	}
	err = vm.MountToolsInstaller(ctx)
	if err != nil {
		return err
	}
// todo, wait for some time
running, err := vm.IsToolsRunning(ctx)
	if err != nil {
		return err
	}

**Comment by atc0005** [comment]
@venkathangella 

This looks to be relevant:

- https://pkg.go.dev/github.com/vmware/govmomi/object

I've not tried to upgrade VMware Tools using govmomi, but it looks like you'll need to use `UpgradeTools` after you mount the VMware Tools installer (ISO image).

**Reply by venkathangella** [reply]
@atc0005 

Thanks for the reply. I have tried the same before posting this thread here, when I use upgrade option, it errors out as below,
"Cannot complete operation because VMware Tools is not running in this virtual machine."

I am not sure about the options argument here, I have tried many random things here.

**Reply by atc0005** [reply]
It looks like the govc CLI tool may be worth a look:

- https://github.com/vmware/govmomi/blob/main/govc/vm/guest/tools.go

If you're able to perform a fresh VMware Tools installation using `govc` then you can use it as a reference for how to replicate the same functionality in your app.

**Reply by venkathangella** [reply]
<img width="525" alt="image" src="https://github.com/vmware/govmomi/assets/5706555/915cf8db-d5d4-4c8b-a921-c65966ea09ae">



@atc0005 , I have tried **govc** as well. **Options**  argument in the command line does not any details about how to make the fresh installation work. I have tried various random keywords here as well. But, I hit the same error as above.

**Reply by atc0005** [reply]
> Options argument in the command line does not any details about how to make the fresh installation work.

I forget where I read it, but it sounds like the options argument is for specifying flags accepted by the VMware Tools installer.

**Reply by venkathangella** [reply]
[atc0005](https://github.com/atc0005) I have gone through multiple blogs and docs. I did not find required information about the same and I see there are some open threads around the same and no one answered.

**Reply by atc0005** [reply]
This may prove useful:

- https://docs.vmware.com/en/VMware-Tools/12.3.0/com.vmware.vsphere.vmwaretools.doc/GUID-E45C572D-6448-410F-BFA2-F729F2CDA8AC.html

Assuming that the `options` argument is for those CLI flags, you may be able to install VMware Tools using the provided example.
